tests/database/test_header_db.py

```diff
@@ -325,6 +465,210 @@ def test_gap_tracking(headerdb, genesis_header, written_headers, evolving_gaps):
+                    # no canonical child, move on...
+                    continue
+                else:
+                    # Both parent and child are canonical, so they must be parent/child
+                    assert child.parent_hash == parent.hash
+
+        action = data.draw(st.sampled_from(ActionEnum))
+        if action == ActionEnum.CHECKPOINT:
+            # Will py-evm reject a checkpoint that already exists? Assuming so, we can
+            # just check the non-present ones...
+            checkpoint_index = data.draw(st.sampled_from(list(sorted(missing_indices[CHAIN_A]))))
+            checkpoint = chain_a_headers[checkpoint_index]
+            checkpoint_score = get_score(genesis_header, chain_a_headers[:checkpoint_index + 1])
+            headerdb.persist_checkpoint_header(checkpoint, checkpoint_score)
+
+            # keep track of whether any checkpoints were added, so we eventually
+            # guarantee A as canonical
+            last_checkpoint = checkpoint_index
+
+            missing_indices[CHAIN_A].discard(checkpoint_index)
+
+        elif action == ActionEnum.PERSIST_CHAIN:
+            # choose the series of headers to add
+            valid_extensions = get_valid_extensions(missing_indices)
+            chain_idx, next_chain_start = data.draw(st.sampled_from(valid_extensions))
+            next_chain_len = data.draw(st.integers(min_value=1, max_value=chain_length))
+            chain_range_end = next_chain_start + next_chain_len
+            next_headers = headers[chain_idx][next_chain_start:chain_range_end]
+
+            # persist them to chain
+            try:
+                headerdb.persist_header_chain(next_headers)
+            except ValidationError:
+                if chain_idx == CHAIN_B:
+                    # TODO verify that this was *supposed* to throw a validation error here?
+                    # For now, just treat it as fine, as long as it's not CHAIN_A
+                    continue
+                else:
+                    raise
+            except ParentNotFound:
+                # Reject the kind of header chain that might overwrite a checkpoint
+                assert chain_idx == CHAIN_B, "Only chain A should be checkpointed"
+                # Kind of a weak assertion, but better than nothing:
+                # The "gap before checkpoint" should only happen if the A header isn't missing
+                assert (next_chain_start - 1) in missing_indices[CHAIN_A]
+                # Persist failed, so retry different action
+                continue
+            except CheckpointsMustBeCanonical:
+                assert chain_idx == CHAIN_B, "Only chain B attempts would decanonize checkpoint"
+                # Persist failed, so retry different action
+                continue
+
+            # mark persisted headers as not missing
+            for inserted_index in range(next_chain_start, chain_range_end):
+                missing_indices[chain_idx].discard(inserted_index)
```

Comment:
I suppose this could be:
```suggestion
            missing_indices[chain_idx].difference_update(
                range(next_chain_start, chain_range_end)
            )
```
Can't decide if that's clearer or not.